vignettes/syntax_extractor.Rmd

```diff
@@ -0,0 +1,305 @@
+
+The Syntax Extractor is a tool that helps us dive into the structure of source code by using an annotated format. It works by converting code into a structured XML format using srcML, making it easier to analyze and understand the underlying architecture of the project. By extracting key pieces of information such as class names, functions, or even comments (like documentation), we can generate meaningful data that helps with things like semantic analysis, code comprehension, and more.
+
+Imagine a codebase as a collection of books in a library. Without a catalog, finding the right book would be nearly impossible. Syntax extraction is like building a catalog for a codebase. It allows us to quickly pinpoint where classes, methods, and important comments are located, and organize this data in a structured way.
+
+For Kaiaulu, this is particularly useful as we move toward deeper analysis, such as applying machine learning techniques like word embeddings to the extracted syntax. This allows us to not only understand how the code is written but also how the pieces relate to each other on a semantic level.
+
+In short, syntax extraction gives us a way to “see the big picture” without having to manually dig through every line of code.
+
+## How the Syntax Extractor Works
+
+At its core, the syntax extractor relies on [srcML](https://www.srcml.org/), which is a tool that converts source code into an XML representation. This XML structure gives us a detailed breakdown of the code: classes, functions, variables, and comments are annotated with specific XML tags, allowing us to query and extract what we need.
+
+In this notebook, we’ll walk through the process of setting up the syntax extractor, running the extraction process, and then querying the annotated code for useful information like class names, namespaces, and documentation.
+
+# Project Configuration File
+
+In a project, source code is spread across multiple repositories and folders. To use Kaiaulu’s syntax extraction functions, you need to configure the system with details about where the source code is stored and where the tools (like srcML) are located.
+
+Kaiaulu uses a project configuration file format to specify the paths and settings necessary for syntax extraction. This allows you to manage different projects and their associated codebases efficiently.
+
+Here’s an example of how a project configuration file might look for syntax extraction (this example configuration file makes it so that the syntax extractor will focus on the relevant .java and .xml files in the Maven project, ignoring test and example files):
+
+```
+version_control:
+  log: /path/to/local/maven/repo/.git
+
+filter:
+  keep_filepaths_ending_with:
+    - .java
+    - .xml
+  remove_filepaths_containing:
+    - test
+    - example
+
+tool:
+  srcml:
+    srcml_path: /usr/local/bin/srcml
+    srcml_output_path: ../../analysis/maven/srcml_output.xml
+
+analysis:
+  topics:
+    - class
+    - method
+    - documentation
```

Comment:
this probably will need revisiting later, it is legacy of another thing we were doing with srcML. We can go over this on call